bpf/process/bpf_process_event.h

```diff
@@ -108,42 +114,50 @@ d_unlinked(struct dentry *dentry)
 }
 
 static inline __attribute__((always_inline)) int
-prepend_name(char *bf, char **buffer, int *buflen, const char *name, u32 dlen)
+prepend_name(char *buf, char **bufptr, int *buflen, const char *name, u32 namelen)
 {
 	char slash = '/';
-	u64 buffer_offset = (u64)(*buffer) - (u64)bf;
-
-	// Change dlen (the dentry name length) to fit in the buffer.
-	// We prefer to store the part of it that fits rather that discard it.
-	if (dlen + 1 /* for the slash */ >= *buflen)
-		dlen = *buflen - 1 /* for the slash */ -
-		       1 /* in order to avoid the case to do *buflen == 0 */;
-
-	*buflen -= (dlen + 1);
-	// This will not happen as in the previous if-clause ensures that *buflen will be > 0
-	// Needed to make the verifier happy in older kernels.
-	if (*buflen <= 0)
+	// contains 1 if the buffer is large enough to contain the whole name and a slash prefix
+	bool write_slash = 1;
+
+	u64 buffer_offset = (u64)(*bufptr) - (u64)buf;
+
+	// Change name and namelen to fit in the buffer.
+	// We prefer to store the part of it that fits rather than discard it.
+	if (namelen >= *buflen) {
+		name += namelen - *buflen;
+		namelen = *buflen;
+		write_slash = 0;
+	}
+
+	*buflen -= (namelen + write_slash);
+
+	// This will not happen as buffer_offset cannot be above 256 and namelen is
+	// bound to 255. Needed to make the verifier happy in older kernels.
+	if (namelen + write_slash > buffer_offset)
 		return -ENAMETOOLONG;
 
-	buffer_offset -= (dlen + 1);
+	buffer_offset -= (namelen + write_slash);
 
 	// This will never happen. buffer_offset is the diff of the initial buffer pointer
 	// with the current buffer pointer. This will be at max 256 bytes (similar to the initial
 	// size).
 	// Needed to bound that for probe_read call.
-	if (buffer_offset > PATH_MAP_SIZE - 256)
+	if (buffer_offset >= MAX_BUF_LEN)
 		return -ENAMETOOLONG;
 
-	probe_read(bf + buffer_offset, sizeof(char), &slash);
+	if (write_slash)
+		probe_read(buf + buffer_offset, sizeof(char), &slash);
```

Comment:
Hm, couldn't this be simplified to avoid the probe_read call?

Reply:
This works on `6.5.0`, let's see if it passes tests on older versions.

```diff
diff --git a/bpf/process/bpf_process_event.h b/bpf/process/bpf_process_event.h
index ff145373..b2b46845 100644
--- a/bpf/process/bpf_process_event.h
+++ b/bpf/process/bpf_process_event.h
@@ -116,7 +116,6 @@ d_unlinked(struct dentry *dentry)
 static inline __attribute__((always_inline)) int
 prepend_name(char *buf, char **bufptr, int *buflen, const char *name, u32 namelen)
 {
-       char slash = '/';
        // contains 1 if the buffer is large enough to contain the whole name and a slash prefix
        bool write_slash = 1;
 
@@ -147,7 +146,7 @@ prepend_name(char *buf, char **bufptr, int *buflen, const char *name, u32 namele
                return -ENAMETOOLONG;
 
        if (write_slash)
-               probe_read(buf + buffer_offset, sizeof(char), &slash);
+               buf[buffer_offset] = '/';
 
        // This ensures that namelen is < 256, which is aligned with kernel's max dentry name length
        // that is 255 (https://elixir.bootlin.com/linux/v5.10/source/include/uapi/linux/limits.h#L12).
```